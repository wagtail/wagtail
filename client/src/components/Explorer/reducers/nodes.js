const defaultPageState = {
  isFetching: false,
  isError: false,
  children: {
    items: [],
    count: 0,
  },
  meta: {
    children: {},
  },
};

/**
 * A single page node in the explorer.
 */
const node = (state = defaultPageState, { type, payload }) => {
  switch (type) {
  case 'OPEN_EXPLORER':
    return state || defaultPageState;

  case 'GET_PAGE_SUCCESS':
    return Object.assign({}, state, payload.data, {
      isError: false,
    });

  case 'GET_CHILDREN_START':
    return Object.assign({}, state, {
      isFetching: true,
    });

  case 'GET_CHILDREN_SUCCESS':
    return Object.assign({}, state, {
      isFetching: false,
      isError: false,
      children: {
        items: state.children.items.slice().concat(payload.items.map(item => item.id)),
        count: payload.meta.total_count,
      },
    });

  case 'GET_PAGE_FAILURE':
  case 'GET_CHILDREN_FAILURE':
    return Object.assign({}, state, {
      isFetching: false,
      isError: true,
    });

  default:
    return state;
  }
};

const defaultState = {};

/**
 * Contains all of the page nodes in one object.
 */
export default function nodes(state = defaultState, { type, payload }) {
  switch (type) {

  case 'OPEN_EXPLORER':
  case 'GET_PAGE_SUCCESS':
  case 'GET_CHILDREN_START':
  case 'GET_PAGE_FAILURE':
  case 'GET_CHILDREN_FAILURE':
    return Object.assign({}, state, {
      // Delegate logic to single-node reducer.
      [payload.id]: node(state[payload.id], { type, payload }),
    });

  // eslint-disable-next-line no-case-declarations
  case 'GET_CHILDREN_SUCCESS':
    const newState = Object.assign({}, state, {
      [payload.id]: node(state[payload.id], { type, payload }),
    });

    payload.items.forEach((item) => {
      newState[item.id] = Object.assign({}, defaultPageState, item);
    });

    return newState;

  default:
    return state;
  }
}
