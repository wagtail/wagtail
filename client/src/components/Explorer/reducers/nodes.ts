import { WagtailPageAPI } from '../../../api/admin';
import { OPEN_EXPLORER, CLOSE_EXPLORER } from './explorer';

export interface PageState extends WagtailPageAPI {
  isFetchingChildren: boolean,
  isFetchingTranslations: boolean,
  isError: boolean;
  children: {
    items: any[];
    count: number;
  };
  translations?: Map<string, number>;
}

const defaultPageState: PageState = {
  id: 0,
  isFetchingChildren: false,
  isFetchingTranslations: false,
  isError: false,
  children: {
    items: [],
    count: 0,
  },
  meta: {
    status: {
      status: '',
      live: false,
      has_unpublished_changes: true
    },
    parent: null,
    children: {},
  },
};

interface OpenExplorerAction {
  type: typeof OPEN_EXPLORER;
  payload: {
    id: number;
  }
}

interface CloseExplorerAction {
  type: typeof CLOSE_EXPLORER;
}

export const GET_PAGE_SUCCESS = 'GET_PAGE_SUCCESS';
interface GetPageSuccess {
  type: typeof GET_PAGE_SUCCESS;
  payload: {
    id: number;
    data: WagtailPageAPI;
  };
}

export const GET_CHILDREN_START = 'GET_CHILDREN_START';
interface GetChildrenStart {
  type: typeof GET_CHILDREN_START;
  payload: {
    id: number;
  };
}

export const GET_CHILDREN_SUCCESS = 'GET_CHILDREN_SUCCESS';
interface GetChildrenSuccess {
  type: typeof GET_CHILDREN_SUCCESS;
  payload: {
    id: number;
    meta: {
      /* eslint-disable-next-line camelcase */
      total_count: number;
    };
    items: WagtailPageAPI[];
  };
}

export const GET_TRANSLATIONS_START = 'GET_TRANSLATIONS_START';
interface GetTranslationsStart {
  type: typeof GET_TRANSLATIONS_START;
  payload: {
    id: number;
  };
}

export const GET_TRANSLATIONS_SUCCESS = 'GET_TRANSLATIONS_SUCCESS';
interface GetTranslationsSuccess {
  type: typeof GET_TRANSLATIONS_SUCCESS;
  payload: {
    id: number;
    meta: {
      /* eslint-disable-next-line camelcase */
      total_count: number;
    };
    items: WagtailPageAPI[];
  };
}

export const GET_PAGE_FAILURE = 'GET_PAGE_FAILURE';
interface GetPageFailure {
  type: typeof GET_PAGE_FAILURE;
  payload: {
    id: number;
  };
}

export const GET_CHILDREN_FAILURE = 'GET_CHILDREN_FAILURE';
interface GetChildrenFailure {
  type: typeof GET_CHILDREN_FAILURE;
  payload: {
    id: number;
  };
}

export const GET_TRANSLATIONS_FAILURE = 'GET_TRANSLATIONS_FAILURE';
interface GetTranslationsFailure {
  type: typeof GET_TRANSLATIONS_FAILURE;
  payload: {
    id: number;
  };
}

export type Action = OpenExplorerAction
                   | CloseExplorerAction
                   | GetPageSuccess
                   | GetChildrenStart
                   | GetChildrenSuccess
                   | GetTranslationsStart
                   | GetTranslationsSuccess
                   | GetPageFailure
                   | GetChildrenFailure
                   | GetTranslationsFailure;

/**
 * A single page node in the explorer.
 */
const node = (state = defaultPageState, action: Action): PageState => {
  switch (action.type) {
  case GET_PAGE_SUCCESS:
    return Object.assign({}, state, action.payload.data, {
      isError: false,
    });

  case GET_CHILDREN_START:
    return Object.assign({}, state, {
      isFetchingChildren: true,
    });

  case GET_TRANSLATIONS_START:
    return Object.assign({}, state, {
      isFetchingTranslations: true,
    });

  case GET_CHILDREN_SUCCESS:
    return Object.assign({}, state, {
      isFetchingChildren: false,
      isError: false,
      children: {
        items: state.children.items.slice().concat(action.payload.items.map(item => item.id)),
        count: action.payload.meta.total_count,
      },
    });

  case GET_TRANSLATIONS_SUCCESS:
    // eslint-disable-next-line no-case-declarations
    const translations = new Map();

    action.payload.items.forEach(item => {
      translations.set(item.meta.locale, item.id);
    });

    return Object.assign({}, state, {
      isFetchingTranslations: false,
      isError: false,
      translations,
    });

  case GET_PAGE_FAILURE:
  case GET_CHILDREN_FAILURE:
  case GET_TRANSLATIONS_FAILURE:
    return Object.assign({}, state, {
      isFetchingChildren: false,
      isFetchingTranslations: true,
      isError: true,
    });

  default:
    return state;
  }
};

export interface State {
  [id: number]: PageState;
}

const defaultState: State = {};

/**
 * Contains all of the page nodes in one object.
 */
export default function nodes(state = defaultState, action: Action) {
  switch (action.type) {
  case OPEN_EXPLORER: {
    return Object.assign({}, state, {
      [action.payload.id]: Object.assign({}, defaultPageState),
    });
  }

  case GET_PAGE_SUCCESS:
  case GET_CHILDREN_START:
  case GET_TRANSLATIONS_START:
  case GET_PAGE_FAILURE:
  case GET_CHILDREN_FAILURE:
  case GET_TRANSLATIONS_FAILURE:
    return Object.assign({}, state, {
      // Delegate logic to single-node reducer.
      [action.payload.id]: node(state[action.payload.id], action),
    });

  case GET_CHILDREN_SUCCESS:
  case GET_TRANSLATIONS_SUCCESS:
    // eslint-disable-next-line no-case-declarations
    const newState = Object.assign({}, state, {
      [action.payload.id]: node(state[action.payload.id], action),
    });

    action.payload.items.forEach((item) => {
      newState[item.id] = Object.assign({}, defaultPageState, item);
    });

    return newState;

  case CLOSE_EXPLORER: {
    return defaultState;
  }

  default:
    return state;
  }
}
