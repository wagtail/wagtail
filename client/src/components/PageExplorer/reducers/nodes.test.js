import nodes from './nodes';

describe('nodes', () => {
  const initialState = nodes(undefined, {});

  it('exists', () => {
    expect(nodes).toBeDefined();
  });

  it('empty state', () => {
    expect(initialState).toMatchSnapshot();
  });

  it('OPEN_EXPLORER', () => {
    const action = { type: 'OPEN_EXPLORER', payload: { id: 1 } };
    expect(nodes(initialState, action)).toMatchSnapshot();
  });

  it('GET_PAGE_SUCCESS', () => {
    const action = { type: 'GET_PAGE_SUCCESS', payload: { id: 1, data: {} } };
    expect(nodes(initialState, action)).toMatchSnapshot();
  });

  it('GET_PAGE_FAILURE', () => {
    const state = nodes(initialState, { type: 'OPEN_EXPLORER', payload: { id: 1 } });
    const action = { type: 'GET_PAGE_FAILURE', payload: { id: 1 } };
    expect(nodes(state, action)).toMatchSnapshot();
  });

  it('GET_CHILDREN_START', () => {
    const action = { type: 'GET_CHILDREN_START', payload: { id: 1 } };
    expect(nodes(initialState, action)).toMatchSnapshot();
  });

  it('GET_CHILDREN_SUCCESS', () => {
    const state = nodes(initialState, { type: 'OPEN_EXPLORER', payload: { id: 1 } });
    const action = {
      type: 'GET_CHILDREN_SUCCESS',
      payload: {
        id: 1,
        items: [
          { id: 3 },
          { id: 4 },
          { id: 5 },
        ],
        meta: {
          total_count: 3,
        },
      },
    };
    expect(nodes(state, action)).toMatchSnapshot();
  });

  it('GET_CHILDREN_FAILURE', () => {
    const state = nodes(initialState, { type: 'OPEN_EXPLORER', payload: { id: 1 } });
    const action = { type: 'GET_CHILDREN_FAILURE', payload: { id: 1 } };
    expect(nodes(state, action)).toMatchSnapshot();
  });
});
