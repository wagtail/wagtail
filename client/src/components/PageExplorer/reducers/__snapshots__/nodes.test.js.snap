// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodes GET_CHILDREN_FAILURE 1`] = `
{
  "1": {
    "children": {
      "count": 0,
      "items": [],
    },
    "id": 0,
    "isError": true,
    "isFetchingChildren": false,
    "isFetchingTranslations": true,
    "meta": {
      "children": {},
      "parent": null,
      "status": {
        "has_unpublished_changes": true,
        "live": false,
        "status": "",
      },
    },
  },
}
`;

exports[`nodes GET_CHILDREN_START 1`] = `
{
  "1": {
    "children": {
      "count": 0,
      "items": [],
    },
    "id": 0,
    "isError": false,
    "isFetchingChildren": true,
    "isFetchingTranslations": false,
    "meta": {
      "children": {},
      "parent": null,
      "status": {
        "has_unpublished_changes": true,
        "live": false,
        "status": "",
      },
    },
  },
}
`;

exports[`nodes GET_CHILDREN_SUCCESS 1`] = `
{
  "1": {
    "children": {
      "count": 3,
      "items": [
        3,
        4,
        5,
      ],
    },
    "id": 0,
    "isError": false,
    "isFetchingChildren": false,
    "isFetchingTranslations": false,
    "meta": {
      "children": {},
      "parent": null,
      "status": {
        "has_unpublished_changes": true,
        "live": false,
        "status": "",
      },
    },
  },
  "3": {
    "children": {
      "count": 0,
      "items": [],
    },
    "id": 3,
    "isError": false,
    "isFetchingChildren": false,
    "isFetchingTranslations": false,
    "meta": {
      "children": {},
      "parent": null,
      "status": {
        "has_unpublished_changes": true,
        "live": false,
        "status": "",
      },
    },
  },
  "4": {
    "children": {
      "count": 0,
      "items": [],
    },
    "id": 4,
    "isError": false,
    "isFetchingChildren": false,
    "isFetchingTranslations": false,
    "meta": {
      "children": {},
      "parent": null,
      "status": {
        "has_unpublished_changes": true,
        "live": false,
        "status": "",
      },
    },
  },
  "5": {
    "children": {
      "count": 0,
      "items": [],
    },
    "id": 5,
    "isError": false,
    "isFetchingChildren": false,
    "isFetchingTranslations": false,
    "meta": {
      "children": {},
      "parent": null,
      "status": {
        "has_unpublished_changes": true,
        "live": false,
        "status": "",
      },
    },
  },
}
`;

exports[`nodes GET_PAGE_FAILURE 1`] = `
{
  "1": {
    "children": {
      "count": 0,
      "items": [],
    },
    "id": 0,
    "isError": true,
    "isFetchingChildren": false,
    "isFetchingTranslations": true,
    "meta": {
      "children": {},
      "parent": null,
      "status": {
        "has_unpublished_changes": true,
        "live": false,
        "status": "",
      },
    },
  },
}
`;

exports[`nodes GET_PAGE_SUCCESS 1`] = `
{
  "1": {
    "children": {
      "count": 0,
      "items": [],
    },
    "id": 0,
    "isError": false,
    "isFetchingChildren": false,
    "isFetchingTranslations": false,
    "meta": {
      "children": {},
      "parent": null,
      "status": {
        "has_unpublished_changes": true,
        "live": false,
        "status": "",
      },
    },
  },
}
`;

exports[`nodes OPEN_EXPLORER 1`] = `
{
  "1": {
    "children": {
      "count": 0,
      "items": [],
    },
    "id": 0,
    "isError": false,
    "isFetchingChildren": false,
    "isFetchingTranslations": false,
    "meta": {
      "children": {},
      "parent": null,
      "status": {
        "has_unpublished_changes": true,
        "live": false,
        "status": "",
      },
    },
  },
}
`;

exports[`nodes empty state 1`] = `{}`;
